src/librustc_ast_lowering/pat.rs

```diff
@@ -128,7 +127,6 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                             Applicability::MaybeIncorrect,
                         )
                         .emit();
-                    break;
```

Comment:
This specific piece of code was conceptually correct – interpreting this pattern as a simple `PatKind::Rest` makes  the most sense.

In that respect, failing to resolve the variable (i.e. using the first iteration of this PR) seems like a better workaround.

The ideal fix would probably be something that ensures users cannot interpret `ddpos` as anything but `PatKind::Rest`. Achieving this might involve reworking the AST representation or something along those lines. Or perhaps something along the lines of the following hack might work:

```rust
            elems.push(self.lower_pat(Pat {
                kind: PatKind::Rest,
                ..pat
            }));
            break;
```

so that the downstream users _also_ end up seeing `..` rather than `x@..` with a potential to mis-interpret it in some way.

Reply:
@nagisa, thanks! I had the same idea, but in this very case that wouldn't help, because we're already past name resolution.